### ISSUE TITLE
unbonding validators may make past rewards unclaimable

### ISSUE THREAD

**Issue by joshua-mir** [paritytech/substrate#6230]
if unbonding period < payout expiry in staking, this check:
https://github.com/paritytech/substrate/blob/82f254e453313cbd39ab2e8fb02eeb35c87fa971/frame/staking/src/lib.rs#L2435
means that past rewards are inaccessible to nominators after their validator `chill`s.

**Comment by bkchr** [comment]
CC @kianenigma

**Comment by kianenigma** [comment]
(noted, will check it out soon and report back)

**Comment by kianenigma** [comment]
> means that past rewards are inaccessible to nominators after their validator chills.

Either I don't get this, or this has nothing to do with chill as far as I can see. Chill will only make your status to be chilled, but you will still be bonded. Hence, the `Self::bonded(&validator_stash)` will indeed still return your controller and everything is good. 

--- 
But there is a similar problem. If you `unbond`, and then call `withdraw_unbonded`, and if your remaining bonded amount is less than ED (or zero), then you get a call to `kill_stash` for your account, which removed the `Bonded` storage and then claiming rewards will no longer work. 

What we could do: Instead of checking that a stash is a validator _now_, we could check that it had an exposure _at that given era_, i.e. 

```diff
- let controller = Self::bonded(&validator_stash).ok_or(Error::<T>::NotStash)?;
+ if <ErasStakers<T>>::exists((era, validator_stash)) { .. } 
```

This means that even if a validator is now unbonded and withdrawn entirely, the previous eras can be claimed. But the problem is that now we don't know the controller of the bonded account at that era, nor do we have the Ledger entry. Saving them per era would be a lot of hassle for a small gain. 

I vote for this to be a `wontfix` at the time being, and we can add a configuration test to ensure `History_dpeth < Bonding duration`, so that even if a validator decides to unbond everything overnight, nominators will have theoretically enough time to claim rewards themselves before the validator can call `withdraw_unbonded`.

--- 

Nonetheless, this raises a question if generally the rewards system has been well tested when placed next to unbonding or not. Is my description above correct? WDYT @shawntabrizi @thiolliere

**Comment by thiolliere** [comment]
Yes it is correct to me, note that none of the reward will have been claimed, so indeed stakers have `min(HistoryDepth, BondingDuration when validator do unbound)` to claim their reward.

currently in polkadot history_depth is 84 eras and bounding duration is 28 eras.

I don't see easy fix:
* checking that there is no unclaimed point in withdraw_unbond would cost a lot
* add the unclaimed era in another storage (when validator witthdraw_unbond) and in do_payout_staker we would check if ledger exists or if the storage give information that the validator hasn't claimed the era yet.
   storage would be: `Vec<ValidatorStashAccountId, unclaimed_rewards: Vec<EraIndex>` and pruned on access or something else.
  this makes code a bit more difficult but doable.

**Comment by kianenigma** [comment]
I am much more in favour of offloading the load to the validator rather than having even more bookkeeping in staking. 

> checking that there is no unclaimed point in withdraw_unbond would cost a lot

How about if it incorporated in the weight and the bailing would pay for it? If you call `withdraw_unbonded` you automatically do the last payments remaining before everything is cleaned.

**Comment by thiolliere** [comment]
> > checking that there is no unclaimed point in withdraw_unbond would cost a lot
> 
> How about if it incorporated in the weight and the bailing would pay for it? If you call `withdraw_unbonded` you automatically do the last payments remaining before everything is cleaned.

* withdraw_unbonded doesn't make any distrinction between validator and nominator so the cost is for both. (and as long as staker can child/nominate/validate we can't distringuish it).
* the minimum cost is: `$history_depth * (Db::read + decode of EraRewardPoints)`, can be fine maybe.

**Comment by kianenigma** [comment]
Potentially we could add an argument to the call where to origin specifies if they are either of 

```rust 
enum Unbonding {
   Validator,
   Nominator,
}
```

when unbonding. 

Let's wait on Shawn's opinion as well. I think this is fine and can be fixed like this.

**Comment by shawntabrizi** [comment]
I dont get this issue. Would things be trivially solved if the validator/nominator made sure to do all their payouts before unbonding?

**Comment by kianenigma** [comment]
> I dont get this issue. Would things be trivially solved if the validator/nominator made sure to do all their payouts before unbonding?

Yes, precisely if validators were force to make sure their book is clean before calling withdraw_unbonded.

**Comment by shawntabrizi** [comment]
This does not sound like an issue that the runtime needs to solve. We do nothing to stop a user from shooting themselves in the foot or from a user not retrieving their funds or whatever.

If a user chooses to unbond, they should make sure they claim whatever outstanding payments are available to them. UI can help here of course.

**Comment by kianenigma** [comment]
Makes sense, i.e. a UI/tool can tell you that your validator has unbonded and you should do something about it.

**Comment by joshua-mir** [comment]
what can a nominator do if their validator unbonds maliciously? If they only check once per payout period their validators could all run away and burn all of their rewards?

**Comment by shawntabrizi** [comment]
What can a nominator do if the validator does any malicious activity?

Nominating a validator is a sign of trust.
